### ISSUE TITLE
Grails 6.2.2: ClassCastException for inheritance in Commands

### ISSUE THREAD

**Issue by agre1981** [grails/grails-core#13945]
### Expected Behavior

Grails throws exception ClassCastException when we have 2 endpoints with inherited commands. It worked fine on 6.1.2.

Looks like this is a critical issue.

```
class UrlMappings {
        "/my/$patientId/procedure"(controller: "my", action: "createLedgerProcedure", method: "POST")
        "/my/$patientId/procedure/insuranceEstimates"(controller: "my", action: "getInsuranceEstimates", method: "POST")
}

class MyBaseCommand implements Validateable {

    int testId
}

class MyCommand extends MyBaseCommand {

    int myId
}

class MyController {
    def getInsuranceEstimates(MyCommand command, long patientId) {
        log.info('test2 method')
        render([response: "test2 method"] as JSON)
    }

    def createLedgerProcedure(MyBaseCommand command, long patientId) {
        log.info('test1 method')
        render([response: "test1 method"] as JSON)
    }

}
```

### Actual Behaviour

Exception:
```
2024-12-26 18:13:10.639 ERROR --- [nio-8080-exec-2] StackTrace                               : Full Stack Trace:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8)
	... 13 common frames omitted

2024-12-26 18:13:10.690 ERROR --- [nio-8080-exec-2] o.g.web.errors.GrailsExceptionResolver   : ClassCastException occurred when processing request: [POST] /my/123/procedure
class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8). Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8)
	... 13 common frames omitted

```

### Steps To Reproduce

Reproduces on Grails 6.2.2. It worked fine on 6.1.2.

[grails_6_2_2.zip](https://github.com/user-attachments/files/18254075/grails_6_2_2.zip)

An example of the application has been attached:
- run the attached application: `./gradlew bootRun`
- Send the request:
```

curl --location --request POST 'http://localhost:8080/my/123/procedure' \
--header 'Content-Type: application/json' \
--data '{
    "testId": 321
}'
```

### Environment Information

- Windows 10
- Java:
```
openjdk version "11.0.22" 2024-01-16
OpenJDK Runtime Environment Temurin-11.0.22+7 (build 11.0.22+7)
OpenJDK 64-Bit Server VM Temurin-11.0.22+7 (build 11.0.22+7, mixed mode)
```

### Version

6.2.2

**Comment by agre1981** [comment]
Maybe relates to https://github.com/grails/grails-core/issues/13486

**Comment by agre1981** [comment]
Class file: 
[MyController.zip](https://github.com/user-attachments/files/18259565/MyController.zip)

**Comment by jdaugherty** [comment]
I added 2 tests for this issue under my 6.2.x branch & 7.x:

    @Issue('https://github.com/grails/grails-core/issues/13945')
    void "calling actions involving inherited command objects - parent command"() {
        given:
        params.name = 'Douglas'

        when:
        def model = controller.methodTakingParent()
        def commandObject = model.commandObject

        then:
        commandObject.name == 'Douglas'
    }

    @Issue('https://github.com/grails/grails-core/issues/13945')
    void "calling actions involving inherited command objects - child command"() {
        given:
        params.name = 'Douglas'
        params.anotherName = 'Fir'

        when:
        def model = controller.methodTakingChild()
        def commandObject = model.commandObject

        then:
        commandObject.name == 'Douglas'
        commandObject.anotherName == 'Fir'
    }

And the associated Commands:

    class ParentCommandObject {
        String name
    }

    class ChildCommandObject extends ParentCommandObject {
        String anotherName
    }

And the associated action:

    def methodTakingParent(ParentCommandObject co) {
        [commandObject: co]
    }

    def methodTakingChild(ChildCommandObject co) {
        [commandObject: co]
    }

Neither of these tests are failing in either branch.  I'm still researching.

**Comment by agre1981** [comment]
@jdaugherty could you push the tests? I'll try to help with reproducing of the issue

**Comment by jdaugherty** [comment]
I figured out the cause of this.  I'll push the tests here shortly.  The name of the method & the arguments to that method seem to matter.  The actual cast exception is happening on the validate() call: 

<img width="1275" alt="image" src="https://github.com/user-attachments/assets/2764ffdf-6a45-4d06-85a4-f7828aa98d9a" />

**Comment by jdaugherty** [comment]
Commit https://github.com/grails/grails-core/commit/268e38664db5ea2b0778db3fb8ea8d354fd1db2f reproduces this issue.  I've gone ahead and marked the test with `@PendingFeature` to not fail the test run until this is fixed.

**Comment by agre1981** [comment]
The test doesn't fail without `Validatable` for me
![image](https://github.com/user-attachments/assets/bed79a25-96f3-4c05-a524-815ca842988b)

**Comment by jdaugherty** [comment]
There seems to be some randomness.  The test run passed unexpectedly so this leads me to believe there's an ordering issue involved.

**Comment by jdaugherty** [comment]
I changed the code that calls validate in the action transformer, and this now is generated: 
<img width="1295" alt="image" src="https://github.com/user-attachments/assets/c7f28b6f-7090-49a0-8377-d3409e23526d" />

Which makes the test pass.

**Comment by agre1981** [comment]
For me, it constantly fails with `Validateable` and constantly passes without `Validateable`. My command:
```
./gradlew :grails-test-suite-web:clean :grails-test-suite-web:singleTest --tests "org.grails.web.commandobjects.CommandObjectsSpec"
```

**Comment by jdaugherty** [comment]
I accidentally pushed the fix for this to the 6.2.x branch, I reverted it there and then opened https://github.com/grails/grails-core/pull/13947 to fix this in 6.2.x.

**Comment by agre1981** [comment]
It can make sense to add `Validatable` to the test to cover my "not failing" case

**Comment by agre1981** [comment]
@jdaugherty Could you advise the tool for decompiling `.class` you use? 

IntelliJ fails to decompile `controller` classes
![image](https://github.com/user-attachments/assets/42328d70-5b1e-4b93-9def-f960a48cf632)

**Comment by matrei** [comment]
Hmm, both tests run successfully for me on 33b3dc1e5c67388fbc4edf3ed1175f9c09adb6e0, without the changes to `ControllerActionTransformer`.

**Comment by agre1981** [comment]
Could you try running the tests with clean?

```
./gradlew :grails-test-suite-web:clean :grails-test-suite-web:singleTest --tests "org.grails.web.commandobjects.CommandObjectsSpec"
```

**Comment by jdaugherty** [comment]
> @jdaugherty Could you advise the tool for decompiling `.class` you use?
> 
> IntelliJ fails to decompile `controller` classes ![image](https://private-user-images.githubusercontent.com/13238053/399316114-42328d70-5b1e-4b93-9def-f960a48cf632.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1NzI2MzYsIm5iZiI6MTczNTU3MjMzNiwicGF0aCI6Ii8xMzIzODA1My8zOTkzMTYxMTQtNDIzMjhkNzAtNWIxZS00YjkzLTlkZWYtZjk2MGE0OGNmNjMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE1MjUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5NTZlODU4NmQ1OTQ1MzU2MjdiYmEwOTU4MzAzZjAxY2U3N2YwYzIyZjMzNjk2ZjA5ZjhiZjcwNTBjZDQ5ZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aetslDyT4YrDcq7rcHpnJcO5_BPeiSat0chSaK7e9-Q)

I use http://java-decompiler.github.io/ to decompile groovy code.

**Comment by agre1981** [comment]
I checked my original project and it doesn't fail without `Validateable` and fails with it. 

And this looks logical because I don't expect calling `validation` for `non-Validateable` commands

**Comment by matrei** [comment]
> Could you try running the tests with clean?

@agre1981 Thanks, I managed to get the tests to fail now.

**Comment by jdaugherty** [comment]
The 6.2.3-SNAPSHOT build is now published.  @agre1981  can you please set your grails version to this version and let me know if it fixes this issue for you?

**Comment by agre1981** [comment]
@jdaugherty  Yes, it works fine on Grails 6.2.3-SNAPSHOT. 

Great job! thank you!

![image](https://github.com/user-attachments/assets/7e87d9a2-0911-421b-b4c4-f1f367d81e70)

**Comment by jdaugherty** [comment]
I'm going to close this ticket as fixed.  I'm discussing with the other maintainers on when best to release this.  Hopefully it's soon.

**Comment by jdaugherty** [comment]
We have released 6.2.3 to address this issue.
